fontforge/parsettf.c

```diff
@@ -4716,31 +4716,44 @@ return;
 	    len = getlong(ttf);
 	    /* language = */ getlong(ttf);
 	}
+	uint16 *dtable = NULL;
+	size_t dlen = len;
+	if (dlen >= 1048576) {
+		// One typeface has over 100,000 entries in a table!
+		IError("Table too large; truncated to 1048576 entries.");
+		dlen = 1048576;
+	}
 	if ( enc->is_unicodebmp && (format==8 || format==10 || format==12))
 	    enc = FindOrMakeEncoding("UnicodeFull");
 
 	if ( format==0 ) {
```

Comment:
If this is truly a format 0 encoded subtable, then it should only have at most 256 entries: https://www.microsoft.com/typography/otspec/cmap.htm#format0

Reply:
In light of that, I'd say this section should remain as before, except for the `memset` fix

Reply:
Eww. Then we're doing something else wrong. I'm attaching a file that set len over 100,000 when I set a breakpoint there. See if I'm missing something.

[fonts-japanese-gothic.zip](https://github.com/fontforge/fontforge/files/1515593/fonts-japanese-gothic.zip)

Reply:
Thanks, I'll see if I can take a look later this evening

Reply:
@frank-trampe  are you sure you attached the right file? I never reach that breakpoint, and all FontForge says is:

```
fonts-japanese-gothic.ttf is not in a known format (or uses features of that format fontforge does not support, or is so badly corrupted as to be unreadable)
```

Reply:
Ohh derp. The field is uint16 so the max is 65535, not 256. So the format 0 table has the length field as being uint16, but the max length can only be 256. So *specifically* for format 0 tables, if the length is > 256 it should be capped to 256.

Reply:
I'm still able to reproduce the big value. I'm using the most recent commit in this branch and breaking right where dlen gets set from len.

```
frank@Luther-1:/var/tmp/ff10/fontforge/fontforgeexe$ gdb ./.libs/lt-fontforge 
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./.libs/lt-fontforge...done.
(gdb) break parsettf.c:4720
No source file named parsettf.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (parsettf.c:4720) pending.
(gdb) run /usr/share/fonts/truetype/fonts-japanese-gothic.ttf
Starting program: /var/tmp/ff10/fontforge/fontforgeexe/.libs/lt-fontforge /usr/share/fonts/truetype/fonts-japanese-gothic.ttf
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Copyright (c) 2000-2014 by George Williams. See AUTHORS for Contributors.
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
 with many parts BSD <http://fontforge.org/license.html>. Please read LICENSE.
 Based on sources from 22:44 UTC 28-Nov-2017-ML-D.
 Based on source from git with hash: 61a4fce33b60e84f0e11e132f4d3243a1b254ae5
Failed to open hotkey definition file: /var/tmp/ff10/fontforge/fontforgeexe/share/fontforge/hotkeys/default
no xdefs_filename!
TESTING: getPixmapDir:/var/tmp/ff10/fontforge/fontforgeexe/share/fontforge/pixmaps
TESTING: getShareDir:/var/tmp/ff10/fontforge/fontforgeexe/share/fontforge
TESTING: GResourceProgramDir:/var/tmp/ff10/fontforge/fontforgeexe/.libs
trying default theme:/var/tmp/ff10/fontforge/fontforgeexe/share/fontforge/pixmaps/resources
Failed to open resource file: /var/tmp/ff10/fontforge/fontforgeexe/share/fontforge/pixmaps/resources
else argv[i]:/usr/share/fonts/truetype/fonts-japanese-gothic.ttf

Breakpoint 1, readttfencodings (ttf=0x7f4200, info=0x7fffffffb3d0, justinuse=0)
    at parsettf.c:4720
4720		size_t dlen = len;
(gdb) p len
$1 = 123616
```

Reply:
As you point out, that's okay since it's not format 0, but it's always eerie when the same code behaves differently in different environments.

Reply:
Looking at the specification, I think that the maximum value for the subtable length, since it includes the three header fields, is actually 256 + 6.

Reply:
Yeah, that is strange. I was able to reproduce an invalid situation with the POC from https://github.com/fontforge/fontforge/issues/3095, but just not with your attached sample.